File: RxTools-library/src/main/java/com/vondear/rxtools/view/popupwindows/tools/RxPopupViewManager.java
package com.vondear.rxtools.view.popupwindows.tools;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.graphics.Outline;
import android.graphics.Point;
import android.os.Build;
import android.support.annotation.NonNull;
import android.util.Log;
import android.view.View;
import android.view.ViewOutlineProvider;
import android.widget.TextView;

import com.vondear.rxtools.RxAnimationTool;

import java.util.HashMap;
import java.util.Map;

/**
 * @author vondear
 */
public class RxPopupViewManager {

    private static final String TAG = RxPopupViewManager.class.getSimpleName();

    private static final int DEFAULT_ANIM_DURATION = 400;

    // Parameter for managing tip creation or reuse
    private Map<Integer, View> mTipsMap = new HashMap<>();

    private int mAnimationDuration;
    private TipListener mListener;

    public RxPopupViewManager() {
        mAnimationDuration = DEFAULT_ANIM_DURATION;
    }

    public RxPopupViewManager(TipListener listener) {
        mAnimationDuration = DEFAULT_ANIM_DURATION;
        mListener = listener;
    }

    public View show(RxPopupView rxPopupView) {
        View tipView = create(rxPopupView);
        if (tipView == null) {
            return null;
        }

        // animate tip visibility
        RxAnimationTool.popup(tipView, mAnimationDuration).start();

        return tipView;
    }

    private View create(RxPopupView rxPopupView) {

        if (rxPopupView.getAnchorView() == null) {
            Log.e(TAG, "Unable to create a tip, anchor view is null");
            return null;
        }

        if (rxPopupView.getRootView() == null) {
            Log.e(TAG, "Unable to create a tip, root layout is null");
            return null;
        }

        // only one tip is allowed near an anchor view at the same time, thus
        // reuse tip if already exist
        if (mTipsMap.containsKey(rxPopupView.getAnchorView().getId())) {
            return mTipsMap.get(rxPopupView.getAnchorView().getId());
        }

        // init tip view parameters
        TextView tipView = createTipView(rxPopupView);

        // on RTL languages replace sides
        if (RxPopupViewTool.isRtl()) {
            switchToolTipSidePosition(rxPopupView);
        }

        // set tool tip background / shape
        RxPopupViewBackgroundConstructor.setBackground(tipView, rxPopupView);

        // add tip to root layout
        rxPopupView.getRootView().addView(tipView);

        // find where to position the tool tip
        Point p = RxPopupViewCoordinatesFinder.getCoordinates(tipView, rxPopupView);

        // move tip view to correct position
        moveTipToCorrectPosition(tipView, p);

        // set dismiss on click
        tipView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss(view, true);
            }
        });

        // bind tipView with anchorView id
        int anchorViewId = rxPopupView.getAnchorView().getId();
        tipView.setTag(anchorViewId);

        // enter tip to map by 'anchorView' id
        mTipsMap.put(anchorViewId, tipView);

        return tipView;

    }

    private void moveTipToCorrectPosition(TextView tipView, Point p) {
        RxCoordinates tipViewRxCoordinates = new RxCoordinates(tipView);
        int translationX = p.x - tipViewRxCoordinates.left;
        int translationY = p.y - tipViewRxCoordinates.top;
        if (!RxPopupViewTool.isRtl()) tipView.setTranslationX(translationX);
        else tipView.setTranslationX(-translationX);
        tipView.setTranslationY(translationY);
    }

    @NonNull
    private TextView createTipView(RxPopupView rxPopupView) {
        TextView tipView = new TextView(rxPopupView.getContext());
        tipView.setTextColor(rxPopupView.getTextColor());
        tipView.setTextSize(rxPopupView.getTextSize());
        tipView.setText(rxPopupView.getMessage() != null ? rxPopupView.getMessage() : rxPopupView.getSpannableMessage());
        tipView.setVisibility(View.INVISIBLE);
        tipView.setGravity(rxPopupView.getTextGravity());
        setTipViewElevation(tipView, rxPopupView);
        return tipView;
    }

    private void setTipViewElevation(TextView tipView, RxPopupView rxPopupView) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            if (rxPopupView.getElevation() > 0) {
                ViewOutlineProvider viewOutlineProvider = new ViewOutlineProvider() {
                    @SuppressLint("NewApi")
                    @Override
                    public void getOutline(View view, Outline outline) {
                        outline.setEmpty();
                    }
                };
                tipView.setOutlineProvider(viewOutlineProvider);
                tipView.setElevation(rxPopupView.getElevation());
            }
        }
    }

    private void switchToolTipSidePosition(RxPopupView rxPopupView) {
        if (rxPopupView.positionedLeftTo()) {
            rxPopupView.setPosition(RxPopupView.POSITION_RIGHT_TO);
        } else if (rxPopupView.positionedRightTo()) {
            rxPopupView.setPosition(RxPopupView.POSITION_LEFT_TO);
        }
    }

    public void setAnimationDuration(int duration) {
        mAnimationDuration = duration;
    }

    public boolean dismiss(View tipView, boolean byUser) {
        if (tipView != null && isVisible(tipView)) {
            int key = (int) tipView.getTag();
            mTipsMap.remove(key);
            animateDismiss(tipView, byUser);
            return true;
        }
        return false;
    }

    public boolean dismiss(Integer key) {
        return mTipsMap.containsKey(key) && dismiss(mTipsMap.get(key), false);
    }

    public View find(Integer key) {
        if (mTipsMap.containsKey(key)) {
            return mTipsMap.get(key);
        }
        return null;
    }

    public boolean findAndDismiss(final View anchorView) {
        View view = find(anchorView.getId());
        return view != null && dismiss(view, false);
    }

    public void clear() {
        if (!mTipsMap.isEmpty()) {
            for (Map.Entry<Integer, View> entry : mTipsMap.entrySet()) {
                dismiss(entry.getValue(), false);
            }
        }
        mTipsMap.clear();
    }

    private void animateDismiss(final View view, final boolean byUser) {
        RxAnimationTool.popout(view, mAnimationDuration, new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                if (mListener != null) {
                    mListener.onTipDismissed(view, (Integer) view.getTag(), byUser);
                }
            }
        }).start();
    }

    public boolean isVisible(View tipView) {
        return tipView.getVisibility() == View.VISIBLE;
    }

    public interface TipListener {
        void onTipDismissed(View view, int anchorViewId, boolean byUser);
    }

}

File: RxUI/src/main/java/com/tamsiree/rxui/view/popupwindows/tools/RxPopupViewManager.kt
package com.tamsiree.rxui.view.popupwindows.tools

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.graphics.Outline
import android.graphics.Point
import android.os.Build
import android.view.View
import android.view.ViewOutlineProvider
import android.widget.TextView
import com.tamsiree.rxkit.RxAnimationTool.popout
import com.tamsiree.rxkit.RxAnimationTool.popup
import com.tamsiree.rxkit.TLog.e
import com.tamsiree.rxui.view.popupwindows.tools.RxPopupViewTool.isRtl
import java.util.*

/**
 * @author tamsiree
 */
class RxPopupViewManager {
    // Parameter for managing tip creation or reuse
    private val mTipsMap: MutableMap<Int, View> = HashMap()
    private var mAnimationDuration: Int
    private var mListener: TipListener? = null

    constructor() {
        mAnimationDuration = DEFAULT_ANIM_DURATION
    }

    constructor(listener: TipListener?) {
        mAnimationDuration = DEFAULT_ANIM_DURATION
        mListener = listener
    }

    fun show(rxPopupView: RxPopupView): View? {
        val tipView = create(rxPopupView) ?: return null

        // animate tip visibility
        popup(tipView, mAnimationDuration.toLong()).start()
        return tipView
    }

    private fun create(rxPopupView: RxPopupView): View? {
        if (rxPopupView.anchorView == null) {
            e(TAG, "Unable to create a tip, anchor view is null")
            return null
        }
        if (rxPopupView.rootView == null) {
            e(TAG, "Unable to create a tip, root layout is null")
            return null
        }

        // only one tip is allowed near an anchor view at the same time, thus
        // reuse tip if already exist
        if (mTipsMap.containsKey(rxPopupView.anchorView.id)) {
            return mTipsMap[rxPopupView.anchorView.id]
        }

        // init tip view parameters
        val tipView = createTipView(rxPopupView)

        // on RTL languages replace sides
        if (isRtl) {
            switchToolTipSidePosition(rxPopupView)
        }

        // set tool tip background / shape
        RxPopupViewBackgroundConstructor.setBackground(tipView, rxPopupView)

        // add tip to root layout
        rxPopupView.rootView.addView(tipView)

        // find where to position the tool tip
        val p = RxPopupViewCoordinatesFinder.getCoordinates(tipView, rxPopupView)

        // move tip view to correct position
        moveTipToCorrectPosition(tipView, p)

        // set dismiss on click
        tipView.setOnClickListener { view -> dismiss(view, true) }

        // bind tipView with anchorView id
        val anchorViewId = rxPopupView.anchorView.id
        tipView.tag = anchorViewId

        // enter tip to map by 'anchorView' id
        mTipsMap[anchorViewId] = tipView
        return tipView
    }

    private fun moveTipToCorrectPosition(tipView: TextView, p: Point) {
        val tipViewRxCoordinates = RxCoordinates(tipView)
        val translationX = p.x - tipViewRxCoordinates.left
        val translationY = p.y - tipViewRxCoordinates.top
        if (!isRtl) tipView.translationX = translationX.toFloat() else tipView.translationX = -translationX.toFloat()
        tipView.translationY = translationY.toFloat()
    }

    private fun createTipView(rxPopupView: RxPopupView): TextView {
        val tipView = TextView(rxPopupView.context)
        tipView.setTextColor(rxPopupView.textColor)
        tipView.textSize = rxPopupView.textSize.toFloat()
        tipView.text = if (rxPopupView.message != null) rxPopupView.message else rxPopupView.spannableMessage
        tipView.visibility = View.INVISIBLE
        tipView.gravity = rxPopupView.textGravity
        setTipViewElevation(tipView, rxPopupView)
        return tipView
    }

    private fun setTipViewElevation(tipView: TextView, rxPopupView: RxPopupView) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            if (rxPopupView.elevation > 0) {
                val viewOutlineProvider: ViewOutlineProvider = object : ViewOutlineProvider() {
                    @SuppressLint("NewApi")
                    override fun getOutline(view: View, outline: Outline) {
                        outline.setEmpty()
                    }
                }
                tipView.outlineProvider = viewOutlineProvider
                tipView.elevation = rxPopupView.elevation
            }
        }
    }

    private fun switchToolTipSidePosition(rxPopupView: RxPopupView) {
        if (rxPopupView.positionedLeftTo()) {
            rxPopupView.position = RxPopupView.POSITION_RIGHT_TO
        } else if (rxPopupView.positionedRightTo()) {
            rxPopupView.position = RxPopupView.POSITION_LEFT_TO
        }
    }

    fun setAnimationDuration(duration: Int) {
        mAnimationDuration = duration
    }

    fun dismiss(tipView: View?, byUser: Boolean): Boolean {
        if (tipView != null && isVisible(tipView)) {
            val key = tipView.tag as Int
            mTipsMap.remove(key)
            animateDismiss(tipView, byUser)
            return true
        }
        return false
    }

    fun dismiss(key: Int): Boolean {
        return mTipsMap.containsKey(key) && dismiss(mTipsMap[key], false)
    }

    fun find(key: Int): View? {
        return if (mTipsMap.containsKey(key)) {
            mTipsMap[key]
        } else null
    }

    fun findAndDismiss(anchorView: View): Boolean {
        val view = find(anchorView.id)
        return view != null && dismiss(view, false)
    }

    fun clear() {
        if (!mTipsMap.isEmpty()) {
            for ((_, value) in mTipsMap) {
                dismiss(value, false)
            }
        }
        mTipsMap.clear()
    }

    private fun animateDismiss(view: View, byUser: Boolean) {
        popout(view, mAnimationDuration.toLong(), object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                super.onAnimationEnd(animation)
                if (mListener != null) {
                    mListener!!.onTipDismissed(view, view.tag as Int, byUser)
                }
            }
        }).start()
    }

    fun isVisible(tipView: View): Boolean {
        return tipView.visibility == View.VISIBLE
    }

    interface TipListener {
        fun onTipDismissed(view: View, anchorViewId: Int, byUser: Boolean)
    }

    companion object {
        private val TAG = RxPopupViewManager::class.java.simpleName
        private const val DEFAULT_ANIM_DURATION = 400
    }
}